cpp/src/arrow/type_traits.h

```diff
@@ -41,102 +41,78 @@ struct TypeTraits<NullType> {
 };
 
 template <>
-struct TypeTraits<UInt8Type> {
-  using ArrayType = UInt8Array;
-  using BuilderType = UInt8Builder;
-  using TensorType = UInt8Tensor;
-  static constexpr int64_t bytes_required(int64_t elements) { return elements; }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return uint8(); }
-};
-
-template <>
-struct TypeTraits<Int8Type> {
-  using ArrayType = Int8Array;
-  using BuilderType = Int8Builder;
-  using TensorType = Int8Tensor;
-  static constexpr int64_t bytes_required(int64_t elements) { return elements; }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return int8(); }
-};
-
-template <>
-struct TypeTraits<UInt16Type> {
-  using ArrayType = UInt16Array;
-  using BuilderType = UInt16Builder;
-  using TensorType = UInt16Tensor;
-
-  static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(uint16_t);
-  }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return uint16(); }
-};
-
-template <>
-struct TypeTraits<Int16Type> {
-  using ArrayType = Int16Array;
-  using BuilderType = Int16Builder;
-  using TensorType = Int16Tensor;
-
-  static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(int16_t);
-  }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return int16(); }
-};
-
-template <>
-struct TypeTraits<UInt32Type> {
-  using ArrayType = UInt32Array;
-  using BuilderType = UInt32Builder;
-  using TensorType = UInt32Tensor;
-
-  static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(uint32_t);
-  }
-  constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return uint32(); }
-};
-
-template <>
-struct TypeTraits<Int32Type> {
-  using ArrayType = Int32Array;
-  using BuilderType = Int32Builder;
-  using TensorType = Int32Tensor;
+struct TypeTraits<BooleanType> {
+  using ArrayType = BooleanArray;
+  using BuilderType = BooleanBuilder;
+  using CType = bool;
 
   static constexpr int64_t bytes_required(int64_t elements) {
-    return elements * sizeof(int32_t);
+    return BitUtil::BytesForBits(elements);
   }
   constexpr static bool is_parameter_free = true;
-  static inline std::shared_ptr<DataType> type_singleton() { return int32(); }
+  static inline std::shared_ptr<DataType> type_singleton() { return boolean(); }
 };
 
 template <>
-struct TypeTraits<UInt64Type> {
-  using ArrayType = UInt64Array;
-  using BuilderType = UInt64Builder;
-  using TensorType = UInt64Tensor;
+struct TypeTraits<bool> {
```

Comment:
Instead of repeating an identical definition, would it be possible to write `struct TypeTraits<bool>: struct TypeTraits<BooleanType> {}`?

Also I find it confusing to use the same template class name for Arrow type traits and C type traits, so perhaps `CTypeTraits<bool>` instead?